app/controllers/concerns/course/forum_topics_concern.rb

```diff
@@ -0,0 +1,35 @@
+module Course::ForumTopicsConcern
+  extend ActiveSupport::Concern
+
+  private
+
+  def authorize_topic_type!(type)
+    case type
+    when 'sticky'.freeze
+      authorize!(:set_sticky, @topic)
+    when 'announcement'.freeze
+      authorize!(:set_announcement, @topic)
+    end
+  end
+
+  def set_lock_state
+    authorize! :edit, @topic
+    @topic.locked = params[:option]
+  end
+
+  def set_hide_state
+    authorize! :edit, @topic
+    @topic.hidden = params[:option]
+  end
+
+  def initial_topic
+    @post = Course::Discussion::Post.new
+    if @topic.save
+      @post = @topic.topic.posts.new(title: topic_params[:title],
+                                     text: post_params[:discussion_post][:text])
+      return true if @post.save
```

Comment:
This can not be done, because I need to use @topic.topic (Discussion::Topic) to build post. But this can not be used before creating Forum::Topic. I'll try to use transaction to create both post and topic.